File: backend/src/routes/posts.ts
import { Router } from "express";
import { getAuth } from "../middlewares/authMiddleware";
import { isAdmin } from "../lib/admin";
import { cachePublic } from "../middlewares/cache";
import { repo } from "../db/repository";
import { ListPostsQueryParams, CreatePostBody } from "../db/schema/zod";
import { paramInt } from "../lib/params";
import { getSubstackPosts } from "../lib/substack";

const router = Router();

router.get("/", cachePublic(60), async (req, res) => {
  try {
    const parsed = ListPostsQueryParams.safeParse(req.query);
    if (!parsed.success) {
      return res.status(400).json({ error: parsed.error.issues });
    }
    const params = parsed.data;
    const { category, tag, limit = 10, offset = 0, authorId } = params;

    const feedUrl = process.env.SUBSTACK_FEED_URL || "https://codedchapter.substack.com/feed";
    if (feedUrl) {
      try {
        let posts = await getSubstackPosts(feedUrl);
        if (category) {
          posts = posts.filter(p => p.category === category);
        }
        if (tag) {
          const lowerTag = tag.toLowerCase();
          posts = posts.filter(p => p.tags.includes(lowerTag));
        }
        return res.json(posts.slice(offset, offset + limit));
      } catch (substackErr) {
        req.log.warn({ err: substackErr }, "Substack feed integration failed, falling back to database");
      }
    }

    const posts = await repo.listPosts(category, tag, limit, offset, authorId);
    return res.json(posts);
  } catch (err: any) {
    req.log.error({ err }, "Failed to list posts");
    return res.status(500).json({
      error: "Internal server error",
      message: err?.message || String(err),
      stack: err?.stack,
    });
  }
});

router.get("/featured", cachePublic(120), async (req, res) => {
  try {
    const feedUrl = process.env.SUBSTACK_FEED_URL || "https://codedchapter.substack.com/feed";
    if (feedUrl) {
      try {
        const posts = await getSubstackPosts(feedUrl);
        return res.json(posts.slice(0, 3));
      } catch (substackErr) {
        req.log.warn({ err: substackErr }, "Substack featured feed integration failed, falling back to database");
      }
    }
    const posts = await repo.getFeaturedPosts();
    res.json(posts);
  } catch (err: any) {
    req.log.error({ err }, "Failed to get featured posts");
    res.status(500).json({
      error: "Internal server error",
      message: err?.message || String(err),
      stack: err?.stack,
    });
  }
});

router.get("/tags", cachePublic(300), async (req, res) => {
  try {
    const feedUrl = process.env.SUBSTACK_FEED_URL || "https://codedchapter.substack.com/feed";
    if (feedUrl) {
      try {
        const posts = await getSubstackPosts(feedUrl);
        const tagsSet = new Set<string>();
        posts.forEach(p => p.tags.forEach(t => tagsSet.add(t)));
        return res.json(Array.from(tagsSet));
      } catch (substackErr) {
        req.log.warn({ err: substackErr }, "Substack tags feed integration failed, falling back to database");
      }
    }
    const tags = await repo.getAllTags();
    res.json(tags);
  } catch (err: any) {
    req.log.error({ err }, "Failed to get tags");
    res.status(500).json({
      error: "Internal server error",
      message: err?.message || String(err),
      stack: err?.stack,
    });
  }
});

router.get("/:id", cachePublic(60), async (req, res): Promise<any> => {
  try {
    const id = paramInt(req.params.id);
    if (isNaN(id)) return res.status(400).json({ error: "Invalid id" });

    const feedUrl = process.env.SUBSTACK_FEED_URL || "https://codedchapter.substack.com/feed";
    if (feedUrl) {
      try {
        const posts = await getSubstackPosts(feedUrl);
        const post = posts.find(p => p.id === id);
        if (post) return res.json(post);
      } catch (substackErr) {
        req.log.warn({ err: substackErr }, "Substack post detail lookup failed, falling back to database");
      }
    }

    const post = await repo.getPost(id);
    if (!post) return res.status(404).json({ error: "Post not found" });
    return res.json(post);
  } catch (err) {
    req.log.error({ err }, "Failed to get post");
    return res.status(500).json({ error: "Internal server error" });
  }
});

router.put("/:id", async (req, res): Promise<any> => {
  try {
    const auth = getAuth(req);
    const userId = auth?.userId;
    if (!userId) return res.status(401).json({ error: "Unauthorized" });

    if (!isAdmin(auth)) {
      return res.status(403).json({ error: "Forbidden: Only the admin can edit chapters." });
    }

    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) return res.status(400).json({ error: "Invalid id" });

    const parsed = CreatePostBody.safeParse(req.body);
    if (!parsed.success) return res.status(400).json({ error: parsed.error });

    const updated = await repo.updatePost(id, userId, parsed.data);
    return res.json(updated);
  } catch (err: any) {
    if (err.message === "Post not found") return res.status(404).json({ error: err.message });
    if (err.message === "Forbidden") return res.status(403).json({ error: err.message });
    req.log.error({ err }, "Failed to update post");
    return res.status(500).json({ error: "Internal server error" });
  }
});

router.delete("/:id", async (req, res): Promise<any> => {
  try {
    const auth = getAuth(req);
    const userId = auth?.userId;
    if (!userId) return res.status(401).json({ error: "Unauthorized" });

    if (!isAdmin(auth)) {
      return res.status(403).json({ error: "Forbidden: Only the admin can delete chapters." });
    }

    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) return res.status(400).json({ error: "Invalid id" });

    const success = await repo.deletePost(id, userId);
    if (!success) return res.status(404).json({ error: "Post not found" });

    return res.json({ success: true });
  } catch (err: any) {
    if (err.message === "Forbidden") return res.status(403).json({ error: err.message });
    req.log.error({ err }, "Failed to delete post");
    return res.status(500).json({ error: "Internal server error" });
  }
});

router.post("/", async (req, res): Promise<any> => {
  try {
    const auth = getAuth(req);
    const userId = auth?.userId;
    if (!userId) return res.status(401).json({ error: "Unauthorized" });

    if (!isAdmin(auth)) {
      return res.status(403).json({ error: "Forbidden: Only the admin can write chapters." });
    }

    const parsed = CreatePostBody.safeParse(req.body);
    if (!parsed.success) return res.status(400).json({ error: parsed.error });

    const authorProfile = await repo.getProfileByUserId(userId);
    const authorName = authorProfile?.displayName || (auth.sessionClaims?.fullName as string) || "Author";
    const created = await repo.createPost(userId, authorName, parsed.data);
    return res.status(201).json(created);
  } catch (err) {
    req.log.error({ err }, "Failed to create post");
    return res.status(500).json({ error: "Internal server error" });
  }
});

export default router;

File: python-backend/routers/posts.py
"""
routers/posts.py
================
Posts API — full CRUD.

DATA SOURCE:
    - USE_SUBSTACK=true  → reads from Substack RSS (current default)
    - USE_SUBSTACK=false → reads from PostgreSQL via services/database.py

All write endpoints (POST, PUT, DELETE) always go to the database.
"""

import logging
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query, Response
from slowapi import Limiter
from slowapi.util import get_remote_address

from config import get_settings, Settings
from middleware.auth import get_current_user, require_auth, require_admin, AuthUser
import services.substack as substack_svc
import services.database as db_svc

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/posts", tags=["Posts"])
limiter = Limiter(key_func=get_remote_address)


def _cache_headers(response: Response, max_age: int) -> None:
    """Apply Cache-Control headers for CDN and browser caching."""
    response.headers["Cache-Control"] = f"public, max-age={max_age}, stale-while-revalidate={max_age * 2}"


# ── GET /api/posts ────────────────────────────────────────────────────────────

@router.get("/", summary="List posts")
async def list_posts(
    response: Response,
    category: Optional[str] = Query(None, description="Filter by 'tech' or 'general'"),
    tag: Optional[str] = Query(None, description="Filter by tag"),
    limit: int = Query(10, ge=1, le=100),
    offset: int = Query(0, ge=0),
    author_id: Optional[str] = Query(None, alias="authorId"),
    settings: Settings = Depends(get_settings),
):
    """
    Returns a paginated list of posts.

    - **Current mode (USE_SUBSTACK=true)**: Fetched from Substack RSS.
    - **Future mode (USE_SUBSTACK=false)**: Fetched from PostgreSQL.
    """
    _cache_headers(response, 60)
    try:
        # ── SUBSTACK MODE ───────────────────────────────────────────────
        # TODO: Remove this block when switching to your own database.
        #       Set USE_SUBSTACK=false in .env to bypass this.
        if settings.use_substack:
            posts = await substack_svc.get_substack_posts(
                settings.substack_feed_url,
                ttl=settings.substack_cache_ttl_seconds,
            )
            # Apply filters manually since RSS has no server-side filtering
            if tag:
                posts = [p for p in posts if tag.lower() in p.tags]
            if category:
                posts = [p for p in posts if p.category == category]
            return [p.model_dump(by_alias=True) for p in posts[offset: offset + limit]]

        # ── DATABASE MODE ───────────────────────────────────────────────
        # This path is active when USE_SUBSTACK=false
        return await db_svc.list_posts(category=category, tag=tag, limit=limit, offset=offset, author_id=author_id)

    except Exception as exc:
        logger.error("list_posts failed: %s", exc, exc_info=True)
        raise HTTPException(status_code=502, detail="Failed to fetch posts")


# ── GET /api/posts/featured ───────────────────────────────────────────────────

@router.get("/featured", summary="Featured posts (latest 3)")
async def get_featured_posts(
    response: Response,
    settings: Settings = Depends(get_settings),
):
    """Returns the 3 most recent posts for the homepage hero."""
    _cache_headers(response, 120)
    try:
        if settings.use_substack:
            posts = await substack_svc.get_substack_posts(
                settings.substack_feed_url,
                ttl=settings.substack_cache_ttl_seconds,
            )
            return [p.model_dump(by_alias=True) for p in posts[:3]]
        return await db_svc.get_featured_posts()
    except Exception as exc:
        logger.error("get_featured_posts failed: %s", exc, exc_info=True)
        raise HTTPException(status_code=502, detail="Failed to fetch featured posts")


# ── GET /api/posts/tags ───────────────────────────────────────────────────────

@router.get("/tags", summary="All unique tags")
async def get_all_tags(
    response: Response,
    settings: Settings = Depends(get_settings),
):
    """Returns all unique tags across all posts."""
    _cache_headers(response, 300)
    try:
        if settings.use_substack:
            posts = await substack_svc.get_substack_posts(
                settings.substack_feed_url,
                ttl=settings.substack_cache_ttl_seconds,
            )
            tag_set: set[str] = set()
            for p in posts:
                tag_set.update(p.tags)
            return sorted(tag_set)
        return await db_svc.get_all_tags()
    except Exception as exc:
        logger.error("get_all_tags failed: %s", exc, exc_info=True)
        raise HTTPException(status_code=502, detail="Failed to fetch tags")


# ── GET /api/posts/{id} ───────────────────────────────────────────────────────

@router.get("/{post_id}", summary="Get single post by ID")
async def get_post(
    post_id: int,
    response: Response,
    settings: Settings = Depends(get_settings),
):
    """Returns a single post by its numeric ID."""
    _cache_headers(response, 60)
    try:
        if settings.use_substack:
            posts = await substack_svc.get_substack_posts(
                settings.substack_feed_url,
                ttl=settings.substack_cache_ttl_seconds,
            )
            post = next((p for p in posts if p.id == post_id), None)
            if not post:
                raise HTTPException(status_code=404, detail="Post not found")
            return post.model_dump(by_alias=True)
        post = await db_svc.get_post(post_id)
        if not post:
            raise HTTPException(status_code=404, detail="Post not found")
        return post
    except HTTPException:
        raise
    except Exception as exc:
        logger.error("get_post(%d) failed: %s", post_id, exc, exc_info=True)
        raise HTTPException(status_code=502, detail="Failed to fetch post")


# ── POST /api/posts ───────────────────────────────────────────────────────────

@router.post("/", status_code=201, summary="Create a post (admin only)")
async def create_post(
    body: dict,
    user: AuthUser = Depends(require_admin),
):
    """
    Create a new post. Admin only.
    Always writes to the database regardless of USE_SUBSTACK setting.
    """
    try:
        return await db_svc.create_post(
            user_id=user.user_id,
            author_name=user.full_name or "CodedChapter",
            data=body,
        )
    except Exception as exc:
        logger.error("create_post failed: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to create post")


# ── PUT /api/posts/{id} ───────────────────────────────────────────────────────

@router.put("/{post_id}", summary="Update a post (admin only)")
async def update_post(
    post_id: int,
    body: dict,
    user: AuthUser = Depends(require_admin),
):
    """Update a post by ID. Admin only."""
    try:
        return await db_svc.update_post(post_id=post_id, user_id=user.user_id, data=body)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc))
    except PermissionError:
        raise HTTPException(status_code=403, detail="Forbidden")
    except Exception as exc:
        logger.error("update_post(%d) failed: %s", post_id, exc, exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to update post")


# ── DELETE /api/posts/{id} ────────────────────────────────────────────────────

@router.delete("/{post_id}", summary="Delete a post (admin only)")
async def delete_post(
    post_id: int,
    user: AuthUser = Depends(require_admin),
):
    """Delete a post and all its comments. Admin only."""
    try:
        deleted = await db_svc.delete_post(post_id=post_id, user_id=user.user_id)
        if not deleted:
            raise HTTPException(status_code=404, detail="Post not found")
        return {"success": True}
    except PermissionError:
        raise HTTPException(status_code=403, detail="Forbidden")
    except HTTPException:
        raise
    except Exception as exc:
        logger.error("delete_post(%d) failed: %s", post_id, exc, exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to delete post")


# ── GET /api/posts/{id}/comments ──────────────────────────────────────────────

@router.get("/{post_id}/comments", summary="List comments on a post")
async def list_comments(post_id: int):
    """Returns all comments for a post, newest first."""
    try:
        return await db_svc.list_comments(post_id)
    except Exception as exc:
        logger.error("list_comments(%d) failed: %s", post_id, exc, exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to fetch comments")


# ── POST /api/posts/{id}/comments ─────────────────────────────────────────────

@router.post("/{post_id}/comments", status_code=201, summary="Add a comment")
async def create_comment(
    post_id: int,
    body: dict,
    user: AuthUser = Depends(require_auth),
):
    """Create a comment on a post. Requires authentication."""
    content = body.get("content", "").strip()
    if not content:
        raise HTTPException(status_code=400, detail="Content is required")
    try:
        return await db_svc.create_comment(
            post_id=post_id,
            author_id=user.user_id,
            author_name=user.full_name or user.email.split("@")[0],
            content=content,
        )
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc))
    except Exception as exc:
        logger.error("create_comment failed: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to create comment")


# ── DELETE /api/posts/{id}/comments/{cid} ─────────────────────────────────────

@router.delete("/{post_id}/comments/{comment_id}", summary="Delete a comment")
async def delete_comment(
    post_id: int,
    comment_id: int,
    user: AuthUser = Depends(require_auth),
):
    """Delete your own comment. Requires authentication."""
    try:
        deleted = await db_svc.delete_comment(comment_id=comment_id, user_id=user.user_id)
        if not deleted:
            raise HTTPException(status_code=404, detail="Comment not found")
        return {"success": True}
    except PermissionError:
        raise HTTPException(status_code=403, detail="Forbidden")
    except HTTPException:
        raise
    except Exception as exc:
        logger.error("delete_comment failed: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail="Failed to delete comment")
